lib/jpegstream.js

```diff
@@ -30,33 +31,39 @@ var Stream = require('stream').Stream;
  */
 
 var JPEGStream = module.exports = function JPEGStream(canvas, options, sync) {
-  var self = this
-    , method = sync
+  Readable.call(this);
+
+  var self = this;
+  var method = sync
       ? 'streamJPEGSync'
       : 'streamJPEG';
   this.options = options;
   this.sync = sync;
   this.canvas = canvas;
-  this.readable = true;
+
   // TODO: implement async
   if ('streamJPEG' == method) method = 'streamJPEGSync';
-  process.nextTick(function(){
-    canvas[method](options.bufsize, options.quality, options.progressive, function(err, chunk){
-      if (err) {
-        self.emit('error', err);
-        self.readable = false;
-      } else if (chunk) {
-        self.emit('data', chunk);
-      } else {
-        self.emit('end');
-        self.readable = false;
-      }
-    });
-  });
-};
 
-/**
- * Inherit from `EventEmitter`.
- */
+  // For now we're not controlling the c++ code's data emission, so we only
+  // call canvas.streamPNGSync once and let it emit data at will.
+  var hasStarted = false;
+
+  self._read = function () {
+    if (!hasStarted) {
+      hasStarted = true;
+      process.nextTick(function(){
+        canvas[method](options.bufsize, options.quality, options.progressive, function(err, chunk){
+            if (err) {
+            self.emit('error', err);
+            } else if (chunk) {
+            self.push(chunk);
+            } else {
+            self.push(null);
+            }
+        });
```

Comment:
I'm confused what's going on here. Evidently... if I store a reference to a C++ function (i.e. `var x = canvas[method]; x();`) then I get an `Illegal invocation` error when I go to invoke it. A simple `canvas[method]()` works though.

In any case, @mcollina I'm curious what the motivation would be for pulling the two functions out into the constructor or outside. Because of the wrapping `if (!hasStarted)` block, I don't see any performance reasons.